compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+  ```
+  To ensure this happens the encoder shrinks the `state` by streaming out its lower bits, *before* encoding each symbol. 
+  This logic is implemented in the function `shrink_state`. Thus, the full encoding step for one symbol is as follows:
+
+  ```python
+    ## Encoding one symbol
+    # output bits to the stream to bring the state in the range for the next encoding
+        state, out_bits = self.shrink_state(state, s)
+        encoded_bitarray = out_bits + encoded_bitarray
+
+    # core encoding step
+    state = self.rans_base_encode_step(s, state)
+  ```
+
+  The decoder does the reverse operation of `expand_state` where, after decoding a symbol, it reads in the a few bits to
+  re-map and expand the state to lie within the acceptable range [L, H]
+  Note that `shrink_state` and `expand_state` are inverses of each other
+
+  Thus, the  full decoding step 
+  ```python
+    # base rANS decoding step
+    s, state = self.rans_base_decode_step(state)
+
+    # remap the state into the acceptable range
+    state, num_bits_used_by_expand_state = self.expand_state(state, encoded_bitarray)
+  ```
+- For completeness: the acceptable range `[L, H]` are given by:
+  L = RANGE_FACTOR*total_freq
+  H = (2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq - 1)
+  Why specifically these values? Look at the *Streaming-rANS* section on https://kedartatwawadi.github.io/post--ANS/
+
+
+## References
+1. Original Asymmetric Numeral Systems paper:  https://arxiv.org/abs/0902.0271
+2. Interactive write-up on rANS: https://kedartatwawadi.github.io/post--ANS/
+
+## TODO
+1. Implement alias-coding style rANS
+2. Adaptive rANS
+"""
+
+from dataclasses import dataclass
+import numpy as np
+from typing import Tuple, Any, List
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from utils.bitarray_utils import BitArray, get_bit_width, uint_to_bitarray, bitarray_to_uint
+from core.data_block import DataBlock
+from core.prob_dist import Frequencies, get_mean_log_prob
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from utils.misc_utils import cache
+
+
+@dataclass
+class rANSParams:
+    """base parameters for the rANS encoder/decoder.
+    More details in the overview
+    """
+
+    # num bits used to represent the data_block size
+    DATA_BLOCK_SIZE_BITS: int = 32
+
+    # the encoder can output NUM_BITS_OUT at a time when it performs the state shrinking operation
+    NUM_BITS_OUT: int = 1  # number of bits
+
+    # rANS state is limited to the range [RANGE_FACTOR*total_freq, (2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq - 1)]
+    # RANGE_FACTOR is a base parameter controlling this range
+    RANGE_FACTOR_BITS: int = 16
+    RANGE_FACTOR: int = 1 << RANGE_FACTOR_BITS
+
+    def initial_state(self, freqs: Frequencies) -> int:
+        """the initial state from which rANS encoding begins
+
+        NOTE: the choice of  this state is somewhat arbitrary, the only condition being, it should lie in the acceptable range
```

Comment:
for arbitrary choices, did we follow some standard reference implementation? Also does it make sense to define L and H as parameters for clarity?

Reply:
I understand that this is just L

Reply:
Yeah, now that i think of it, L, H might be better to be defined beforehand. Also, what do you think about including `freq` in the rANSParams`? that would simplify quite a few things

Reply:
including `freq` in the rANSParams makes sense

Reply:
not doing this now.. ll refactor if and when we implement the adaptive version